Merge v1.5.1 to Stable (#194)
opendatahub-io/data-science-pipelines: proposals/bring_your_own_argo_workflows/byoaw_test_plan.md

> **Assisted-by**: Cursor

Reviewer: Do we really want to leave this here?

Reviewer 2: I am happy to take it out but that's what the AI usage guidelines say

Reviewer 2: @hbelmiro not sure if we came to a conclusion on this? should I leave it as it is ?

# Test Plan: Bring Your Own Argo Workflows (BYOAW)

## Table of Contents
1. [Overview](#overview)
2. [Test Scope](#test-scope)
3. [Test Environment Requirements](#test-environment-requirements)
4. [Test Categories](#test-categories)

> 5. [Test Execution Schedule](#test-execution-schedule)

Reviewer: Broken link.

Reviewer 2: removed

6. [Success Criteria](#success-criteria)
7. [Risk Assessment](#risk-assessment)

## Overview

This test plan validates the "Bring Your Own Argo Workflows" feature, which enables Data Science Pipelines to work with existing Argo Workflows installations instead of deploying dedicated WorkflowControllers. The feature includes a global configuration mechanism to disable DSP-managed WorkflowControllers and ensures compatibility with user-provided Argo Workflows.

The plan covers comprehensive testing scenarios including:
- **Co-existence validation** of DSP and external Argo controllers competing for same events
- **Pre-existing Argo detection** and prevention mechanisms
- **CRD update-in-place** functionality and conflict resolution
- **RBAC compatibility** across different permission models (cluster vs namespace level)
- **Workflow schema version compatibility** and API compatibility validation
- **Z-stream (patch) version compatibility** testing
- **Data preservation** for WorkflowTemplates, CronWorkflows, and pipeline data

> - **Independent lifecycle management** of RHOAI and external Argo installations

Reviewer: ```suggestion
- **Independent lifecycle management** of ODH and external Argo Workflows installations
```

- **Project-level access controls** ensuring workflow visibility boundaries
- **Comprehensive migration scenarios** and upgrade path validation

## Test Scope

### In Scope
- Global configuration toggle to disable/enable WorkflowControllers across all DSPAs
- Compatibility validation with external Argo Workflows installations
- Version compatibility matrix testing (N and N-1 versions)
- Migration scenarios between DSP-managed and external Argo configurations
- Conflict detection and resolution mechanisms
- Co-existence testing of DSP and external WorkflowControllers competing for same events
- RBAC compatibility across different permission models (cluster vs namespace level)
- Workflow schema version compatibility validation
- DSPA lifecycle management with external Argo
- Security and RBAC integration with external Argo
- Performance impact assessment

> - Upgrade scenarios for RHOAI with external Argo

Reviewer: ```suggestion
- Upgrade scenarios for ODH with external Argo
```

- Hello world pipeline validation in co-existence scenarios

### Out of Scope
- Partial ArgoWF installs combined with DSP-shipped Workflow Controller
- Isolation between DSP ArgoWF WC and vanilla cluster-scoped ArgoWF installation

## Test Environment Requirements

### Prerequisites

> - OpenShift/Kubernetes clusters with RHOAI/DSP installed

Reviewer: ```suggestion
- OpenShift/Kubernetes clusters with ODH/DSP installed
```

> - OpenShift/Kubernetes clusters with RHOAI/DSP installed

Reviewer: Is `Kubernetes` mentioned here because we may want to use Kind instead of a real OpenShift cluster?

Reviewer 2: Yup, but I can take it out if that's confusing as this feature is currently targeted for Openshift only, but we may test it on a kind cluster as well

Reviewer: I don't have a strong opinion.

Reviewer 2: I'll leave it as it is in that case :)

- Multiple test environments with different Argo Workflows versions
- Access to modify DataScienceCluster and DSPA configurations
- Sample pipelines covering various complexity levels
- Test data for migration scenarios

### Test Environments
| Environment | Argo Version | DSP Version | Purpose |
|-------------|--------------|-------------|---------|
| Env-1 | 3.4.16 | Current | N version compatibility |
| Env-2 | 3.4.15 | Current | N-1 version compatibility |
| Env-3 | 3.5.x | Current | Future compatibility testing |
| Env-4 | Various | Previous | Upgrade scenarios |

## Test Categories

## 1. Cluster Configuration Tests

### 1.1 Global Configuration Toggle

| Test Case ID | TC-CC-001 |
|---|---|
| **Test Case Summary** | Verify global toggle to disable WorkflowControllers works correctly |

> | **Test Steps** | 1. Install RHOAI with default configuration (WorkflowControllers enabled)<br/>2. Create DSPA and verify WorkflowController deployment<br/>3. Update DataScienceCluster to disable WorkflowControllers:<br/>`spec.components.datasciencepipelines.argoWorkflowsControllers.managementState: Removed`<br/>4. Verify existing WorkflowControllers are removed<br/>5. Create new DSPA and verify no WorkflowController is deployed |

Reviewer: ```suggestion
| **Test Steps** | 1. Install ODH with default configuration (WorkflowControllers enabled)<br/>2. Create DSPA and verify WorkflowController deployment<br/>3. Update DataScienceCluster to disable WorkflowControllers:<br/>`spec.components.datasciencepipelines.argoWorkflowsControllers.managementState: Removed`<br/>4. Verify existing WorkflowControllers are removed<br/>5. Create new DSPA and verify no WorkflowController is deployed |
```

| Test Case ID | TC-CC-002 |
|---|---|
| **Test Case Summary** | Verify re-enabling WorkflowControllers after global disable |
| **Test Steps** | 1. Start with globally disabled WorkflowControllers<br/>2. Create DSPA without WorkflowController<br/>3. Re-enable WorkflowControllers globally<br/>4. Verify WorkflowController is deployed to existing DSPA<br/>5. Create new DSPA and verify WorkflowController deployment |
| **Expected Results** | - Global re-enable successfully restores WorkflowController deployment<br/>- Existing DSPAs receive WorkflowControllers<br/>- New DSPAs deploy with WorkflowControllers<br/>- Pipeline history and data preserved |

### 1.2 Kubernetes Native Mode

| Test Case ID | TC-CC-003 |
|---|---|
| **Test Case Summary** | Verify BYOAW compatibility with Kubernetes Native Mode |
| **Test Steps** | 1. Configure cluster for Kubernetes Native Mode<br/>2. Install external Argo Workflows<br/>3. Disable DSP WorkflowControllers globally<br/>4. Create DSPA and execute pipelines<br/>5. Verify Kubernetes native execution with external Argo |

> | **Expected Results** | - Kubernetes Native Mode works with external Argo<br/>- Pipeline execution uses Kubernetes-native constructs<br/>- No conflicts between modes |

Reviewer: What do you mean by `No conflicts between modes`?
Can you elaborate?

Reviewer 2: I'll rephrase it, now that you point it out, it is confusing :thinking:

Reviewer 2: rephrased the confusing text, thanks for pointing it out

| **Expected Results** | - Platform-level CRDs remain intact<br/>- User CRD modifications preserved<br/>- RBAC conflicts resolved without breaking functionality<br/>- Platform operator doesn't overwrite user changes |

### 1.6 Sub-Component Removal Testing

| Test Case ID | TC-CC-007 |
|---|---|
| **Test Case Summary** | Verify sub-component removal functionality for WorkflowControllers |
| **Test Steps** | 1. Deploy DSPA with WorkflowController enabled<br/>2. Execute pipelines and accumulate run data<br/>3. Disable WorkflowController globally<br/>4. Verify WorkflowController is removed but data preserved<br/>5. Verify backing data (run details, metrics) remains intact<br/>6. Test re-enabling WorkflowController preserves historical data |
| **Expected Results** | - WorkflowController removed cleanly<br/>- Run details and metrics preserved<br/>- Historical pipeline data remains accessible<br/>- Re-enabling restores full functionality |

### 1.7 Pre-existing Argo Detection and Prevention

| Test Case ID | TC-CC-008 |
|---|---|
| **Test Case Summary** | Verify detection and prevention of DSPA creation when pre-existing Argo exists |

> | **Test Steps** | 1. Install external Argo Workflows on cluster<br/>2. Install RHOAI DSP operator<br/>3. Attempt to create DSPA with default configuration (WC enabled)<br/>4. Verify detection mechanism identifies pre-existing Argo<br/>5. Test prevention of DSPA creation or automatic WC disable<br/>6. Verify appropriate warning/guidance messages<br/>7. Test manual override if supported |

Reviewer: ```suggestion
| **Test Steps** | 1. Install external Argo Workflows on cluster<br/>2. Install ODH DSP operator<br/>3. Attempt to create DSPA with default configuration (WC enabled)<br/>4. Verify detection mechanism identifies pre-existing Argo<br/>5. Test prevention of DSPA creation or automatic WC disable<br/>6. Verify appropriate warning/guidance messages<br/>7. Test manual override if supported |
```

| **Expected Results** | - CRDs updated in-place successfully<br/>- Existing Workflows, WorkflowTemplates, CronWorkflows preserved<br/>- New CRs work with updated schema<br/>- No data loss or corruption<br/>- Compatibility maintained |

## 2. Positive Functional Tests

### 2.1 Basic Pipeline Execution

| Test Case ID | TC-PF-001 |
|---|---|
| **Test Case Summary** | Verify basic pipeline execution with external Argo |
| **Test Steps** | 1. Configure DSPA with external Argo<br/>2. Submit simple addition pipeline<br/>3. Monitor execution through DSP UI<br/>4. Verify completion and results<br/>5. Check logs and artifacts |
| **Expected Results** | - Pipeline submits successfully<br/>- Execution progresses normally<br/>- Results accessible through DSP interface<br/>- Logs and monitoring functional |

> ### 2.2 Complex Pipeline Types

Reviewer: All the tests in this section have the same Test Case Summary.

Reviewer 2: Copy pasta issue :) , I can change that

Reviewer 2: Done

> ### 2.2 Complex Pipeline Types

Reviewer: Is `Certificate handling operates securely` really in the scope of all these tests?

Reviewer 2: Its one of the param for driver, so I wanted to cover all the scenarios (also, what if there is a mismatch, we need to make sure driver and launcher handles that correctly)- we probably do not care about argo controller in this case

> ### 2.2 Complex Pipeline Types

Reviewer: It looks like all of the tests under this section have the same expected results, which seem too broad for most tests.

Reviewer 2: its a copy paste issue, I'll fix it

Reviewer 2: Done

| **Test Steps**        | 1. Execute pipeline - Pipelines with artifacts<br/>2. Verify each pipeline type executes correctly<br/>3. Validate artifacts, metadata, and custom configurations                                                                              |
| **Expected Results**  | - All pipeline types execute successfully<br/>- Custom components and packages work correctly<br/>- Retry and iteration logic functions properly<br/>- Certificate handling operates securely<br/>- Artifacts and metadata preserved correctly |

| Test Case ID          | TC-PF-003                                                                                                                                                                                                                         |
|-----------------------|-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|

> | **Test Case Summary** | Execute comprehensive pipeline types from valid pipeline files                                                                                                                                                                    |

Reviewer: What is a `pipeline type`?

Reviewer 2: Its just me naming it that way but its basically pipelines with different configurations like with/without artifacts, with/without input params etc.

> | **Test Steps**        | 1. Execute Pipeline - Pipelines without artifacts<br/>2. Verify each pipeline type executes correctly<br/>3. Validate artifacts, metadata, and custom configurations                                                              |                                                                                                                                                                                   |

Reviewer: There are no artifacts in this test.

```suggestion
| **Test Steps**        | 1. Execute Pipeline - Pipelines without artifacts<br/>2. Verify each pipeline type executes correctly<br/>3. Validate metadata, and custom configurations                                                              |                                                                                                                                                                                   |
```

> | **Expected Results**  | - Pipeline runs successfully<br/>- Custom components and packages work correctly<br/>- Retry and iteration logic functions properly<br/>- Certificate handling operates securely<br/>- Artifacts and metadata preserved correctly |

Reviewer: Is `Retry and iteration logic functions properly` in the scope of this test?

> | **Expected Results**  | - Pipeline runs successfully<br/>- Custom components and packages work correctly<br/>- Retry and iteration logic functions properly<br/>- Certificate handling operates securely<br/>- Artifacts and metadata preserved correctly |

Reviewer: There are no artifacts in this test.

```suggestion
| **Expected Results**  | - Pipeline runs successfully<br/>- Custom components and packages work correctly<br/>- Retry and iteration logic functions properly<br/>- Certificate handling operates securely<br/>- Metadata preserved correctly |
```

Reviewer 2: good catch, fixed

|-----------------------|-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| **Test Case Summary** | Execute comprehensive pipeline types from valid pipeline files                                                                                                                                                                    |
| **Test Steps**        | 1. Execute Pipeline - Conditional branching pipelines <br/>2. Verify each pipeline type executes correctly<br/>3. Validate artifacts, metadata, and custom configurations                                                         |                                                                                                                                                                                   |
| **Expected Results**  | - Pipeline runs successfully<br/>- Custom components and packages work correctly<br/>- Retry and iteration logic functions properly<br/>- Certificate handling operates securely<br/>- Artifacts and metadata preserved correctly |

> ### 2.3 Pod Spec Override Testing

Reviewer: Same as the previous section regarding using the same summary and expected results for all the tests.

Reviewer 2: Fixing

Reviewer 2: fixed for this section

## 9. Migration and Upgrade Tests

### 9.1 DSP-Managed to External Migration

| Test Case ID | TC-MU-001 |
|---|---|
| **Test Case Summary** | Verify migration from DSP-managed to external Argo |
| **Test Steps** | 1. Create DSPA with internal WorkflowController<br/>2. Execute pipelines and accumulate data<br/>3. Install external Argo<br/>4. Disable internal WCs globally<br/>5. Verify data preservation and new execution |
| **Expected Results** | - Migration completes without data loss<br/>- Historical data remains accessible<br/>- New pipelines use external Argo<br/>- Artifacts and metadata preserved |

### 9.2 External to DSP-Managed Migration

| Test Case ID | TC-MU-002 |
|---|---|
| **Test Case Summary** | Verify migration from external to DSP-managed Argo |
| **Test Steps** | 1. Configure DSPA with external Argo<br/>2. Execute pipelines and verify data<br/>3. Re-enable internal WCs globally<br/>4. Remove external Argo configuration<br/>5. Verify continued operation |
| **Expected Results** | - Migration to internal WC successful<br/>- Pipeline history preserved<br/>- New pipelines use internal WC<br/>- No service interruption |

> ### 9.3 RHOAI Upgrade Scenarios

Reviewer: ```suggestion
### 9.3 ODH Upgrade Scenarios
```

| Test Case ID | TC-MU-003 |
|---|---|

> | **Test Case Summary** | Verify RHOAI upgrade preserves external Argo setup |

Reviewer: ```suggestion
| **Test Case Summary** | Verify ODH upgrade preserves external Argo setup |
```

> | **Test Steps** | 1. Configure RHOAI with external Argo<br/>2. Execute baseline pipeline tests<br/>3. Upgrade RHOAI to newer version<br/>4. Verify external Argo configuration intact<br/>5. Re-execute pipeline tests |

Reviewer: ```suggestion
| **Test Steps** | 1. Configure ODH with external Argo<br/>2. Execute baseline pipeline tests<br/>3. Upgrade ODH to newer version<br/>4. Verify external Argo configuration intact<br/>5. Re-execute pipeline tests |
```

| **Expected Results** | - Upgrade preserves BYOAW configuration<br/>- External Argo continues working<br/>- No functionality regression<br/>- Configuration settings maintained |

### 9.4 Argo Version Upgrade with External Installation

| Test Case ID | TC-MU-004 |
|---|---|
| **Test Case Summary** | Verify external Argo version upgrade scenarios |

> | **Test Steps** | 1. Configure DSPA with external Argo version N-1<br/>2. Execute baseline pipeline tests<br/>3. Upgrade external Argo to version N<br/>4. Verify compatibility matrix adherence<br/>5. Test pipeline execution post-upgrade<br/>6. Document any required RHOAI updates |

Reviewer: ```suggestion
| **Test Steps** | 1. Configure DSPA with external Argo version N-1<br/>2. Execute baseline pipeline tests<br/>3. Upgrade external Argo to version N<br/>4. Verify compatibility matrix adherence<br/>5. Test pipeline execution post-upgrade<br/>6. Document any required ODH updates |
```

> | **Expected Results** | - External Argo upgrade completes successfully<br/>- Compatibility maintained within support matrix<br/>- Clear guidance for required RHOAI updates<br/>- Pipeline functionality preserved |

Reviewer: ```suggestion
| **Expected Results** | - External Argo upgrade completes successfully<br/>- Compatibility maintained within support matrix<br/>- Clear guidance for required ODH updates<br/>- Pipeline functionality preserved |
```

### 9.5 Independent Lifecycle Management

| Test Case ID          | TC-MU-005                                                                                                                                                                                                                                                                                                                                                                                          |
|-----------------------|----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|

> | **Test Case Summary** | Verify independent lifecycle management of RHOAI and external Argo                                                                                                                                                                                                                                                                                                                                 |

Reviewer: ```suggestion
| **Test Case Summary** | Verify independent lifecycle management of ODH and external Argo                                                                                                                                                                                                                                                                                                                                 |
```

> | **Test Steps**        | 1. Install and configure RHOAI with external Argo<br/>2. Perform independent upgrade of external Argo installation<br/>3. Verify RHOAI continues operating without issues<br/>4. Perform independent upgrade of RHOAI<br/>5. Verify external Argo continues operating without issues<br/>6. Test independent scaling of each component<br/>7. Verify independent maintenance and restart scenarios |

Reviewer: ```suggestion
| **Test Steps**        | 1. Install and configure ODH with external Argo<br/>2. Perform independent upgrade of external Argo installation<br/>3. Verify ODH continues operating without issues<br/>4. Perform independent upgrade of ODH<br/>5. Verify external Argo continues operating without issues<br/>6. Test independent scaling of each component<br/>7. Verify independent maintenance and restart scenarios |
```

5. **Migration Documentation** - Complete procedures for all migration scenarios and lifecycle management
6. **Data Preservation Guidelines** - Best practices for maintaining data integrity during all transitions
7. **Uninstall Procedures** - Validated procedures for clean removal preserving WorkflowTemplates and CronWorkflows
8. **CRD Management Guidelines** - Platform-level CRD update-in-place and conflict resolution procedures
9. **Pre-existing Argo Detection Guide** - Implementation and configuration of detection mechanisms
10. **Configuration Validation Guide** - Detection and resolution of all unsupported configurations
11. **RBAC Compatibility Matrix** - Comprehensive guidelines for DSP and external Argo RBAC alignment
12. **Schema Compatibility Guide** - Workflow schema version compatibility and API compatibility matrix
13. **Co-existence Best Practices** - Detailed recommendations for running DSP and external Argo together
14. **Z-Stream Testing Strategy** - Framework for ongoing patch version compatibility validation
15. **API Compatibility Documentation** - DSP API Server to external WorkflowController compatibility guidelines

> 16. **Independent Lifecycle Management Guide** - Best practices for managing RHOAI and Argo independently

Reviewer: ```suggestion
16. **Independent Lifecycle Management Guide** - Best practices for managing ODH and Argo independently
```

- Conflict detection and resolution mechanisms
- Co-existence testing of DSP and external WorkflowControllers competing for same events
- RBAC compatibility across different permission models (cluster vs namespace level)
- Workflow schema version compatibility validation
- DSPA lifecycle management with external Argo
- Security and RBAC integration with external Argo
- Performance impact assessment
- Upgrade scenarios for RHOAI with external Argo
- Hello world pipeline validation in co-existence scenarios

### Out of Scope
- Partial ArgoWF installs combined with DSP-shipped Workflow Controller
- Isolation between DSP ArgoWF WC and vanilla cluster-scoped ArgoWF installation

## Test Environment Requirements

### Prerequisites
- OpenShift/Kubernetes clusters with RHOAI/DSP installed
- Multiple test environments with different Argo Workflows versions
- Access to modify DataScienceCluster and DSPA configurations
- Sample pipelines covering various complexity levels
- Test data for migration scenarios

### Test Environments
| Environment | Argo Version | DSP Version | Purpose                       |
|-------------|--------------|-------------|-------------------------------|

> | Env-1       | 3.4.16       | Current     | N version compatibility       |

Reviewer 3: we should avoid using real Argo versions here or else it could be misunderstood as actual support for these explicit versions.

Reviewer 2: How about I change this to:

### Test Environments
| Environment | Argo Version   | DSP Version | Purpose                       |
|-------------|----------------|-------------|-------------------------------|
| Env-1       | Current(3.7.x) | Current     | N version compatibility       |
| Env-2       | 3.6.x          | Current     | N-1 version compatibility     |
| Env-3       | 3.4.x - 3.5.x  | Previous    | Upgrade scenarios             |

?

Reviewer 3: very minor nitpick but i'd list the argo versions for the upgrade scenario as `3.4.x -> 3.5.y`, to signfy the upgrade between two very specific versions rather than a full range, and also that the z-stream version doesn't have to match between argo versions

Reviewer 2: Fixed

|-----------------------|--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| **Test Case Summary** | Verify global toggle to disable WorkflowControllers works correctly                                                                                                                                                                                                                                                                                                                                                            |

> | **Test Steps**        | 1. Install RHOAI with default configuration (WorkflowControllers enabled)<br/>2. Create DSPA and verify WorkflowController deployment<br/>3. Update DataScienceCluster to disable WorkflowControllers:<br/>`spec.components.datasciencepipelines.argoWorkflowsControllers.managementState: Removed`<br/>4. Verify existing WorkflowControllers are removed<br/>5. Create new DSPA and verify no WorkflowController is deployed |

Reviewer 3: very minor suggestion: for numbered lists we can use `1. step1 <br />1. step2<br />1. step3` for every item and markdown will auto-increment the number when rendered.  For maintainability it might be an idea to use this so that it's easier to add/remove steps without needing to manually re-number the rest of the list items

Reviewer 2: I don't think I understand the point, do you want to change the all numbers in the numbered list to 1? or I can change it to bulleted points as well, if the concern is manually alteration in the future

Reviewer 2: Done, used HTML numbered list pattern

### 1.5 Platform-Level CRD and RBAC Management

| Test Case ID          | TC-CC-006                                                                                                                                                                                                                                                                                                                             |
|-----------------------|---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| **Test Case Summary** | Verify platform-level Argo CRDs and RBAC remain intact with external Argo                                                                                                                                                                                                                                                             |

> | **Test Steps**        | 1. Install DSPO which creates platform-level Argo CRDs and RBAC<br/>2. Install external Argo with different CRD versions<br/>3. Toggle global WorkflowController disable<br/>4. Verify platform CRDs are not removed<br/>5. Test that user modifications to CRDs are preserved<br/>6. Verify RBAC conflicts are handled appropriately |

Reviewer 3: This is more a document-encompassing comment, but should this test plan be in the DSPO repo instead, and/or perhaps we should split this test plan up into 2: one for DSPO testing and one for DSP w/ External Argo (No DSPO used) tests?

Reviewer 3: for context, there are no references to DSPO (except for a couple instances in an internal procedure doc) in this entire repo), so I believe the intent is the DSP repo is 'unaware' of the existence of DSPO

Reviewer 2: :thinking: My thinking with this is that, DSP comes as a package with DSPO in a RHAOI cluster, so it doesn;t matter where this exist, but I see your point. I can move some of these tests to DSPO repo

Reviewer 2: On second thoughts, I think its better to keep the whole test plan in a single place, otherwise we may miss to execute/implement certain portion of it

Reviewer 3: if that is the case, then i think we should probably migrate this to the DSPO repo then, as the base DSP repo shouldn't have any real backreferences to the DSPOperator and/or platform operators (once this is ready to merge)

Reviewer 2: Hmmmm..... good point
I can do that

Reviewer 2: Actually, on second thoughts, as I was about to move this over, I am thinking that since the original proposal also exists here in this repo, we should probably leave the test plan here along with the proposal as well

Reviewer 3: So FWIW I dont believe this repo has any BYOArgo materials in it either, including any proposals.

| **Test Case Summary** | Verify sub-component removal functionality for WorkflowControllers                                                                                                                                                                                                                                                                              |
| **Test Steps**        | 1. Deploy DSPA with WorkflowController enabled<br/>2. Execute pipelines and accumulate run data<br/>3. Disable WorkflowController globally<br/>4. Verify WorkflowController is removed but data preserved<br/>5. Verify backing data (run details, metrics) remains intact<br/>6. Test re-enabling WorkflowController preserves historical data |
| **Expected Results**  | - WorkflowController removed cleanly<br/>- Run details and metrics preserved<br/>- Historical pipeline data remains accessible<br/>- Re-enabling restores full functionality                                                                                                                                                                    |

> ### 1.7 Pre-existing Argo Detection and Prevention

Reviewer 3: This is currently a P1 requirement, dependent on the ability to do argo detection by the DSPO.  For now, this mechanism isn't yet implemented, and so this would be a failing test.  Given that, we should somehow mark this as optional (for now), or remove it until the capability is implemented

Reviewer 2: We can add a section in this doc where we can define phases and link tickets. The intent of this doc is show coverage for all areas associated with the changes, and if there are things that are going to be implemented in the future, then we can mark those accordingly.

Reviewer 2: Added a section, but I need your help in adding the test cases to that section based on the feature ticket implementation plan

| **Test Steps**        | 1. Install external Argo with specific CRD version<br/>2. Create Workflows, WorkflowTemplates, and CronWorkflows<br/>3. Install DSP with different compatible CRD version<br/>4. Verify CRDs are updated in-place<br/>5. Verify existing CRs (Workflows, WorkflowTemplates, CronWorkflows) remain intact<br/>6. Test new CR creation with updated CRD schema<br/>7. Verify no data loss or corruption |
| **Expected Results**  | - CRDs updated in-place successfully<br/>- Existing Workflows, WorkflowTemplates, CronWorkflows preserved<br/>- New CRs work with updated schema<br/>- No data loss or corruption<br/>- Compatibility maintained                                                                                                                                                                                      |

> ## 2. Positive Functional Tests

Reviewer 3: We should include a summary of _why_ we're doing these tests - the context being to validate Argo functionality is not broken in what we've marked as 'compatible' versions of Argo.  As it stands now, these look like generic blackbox DSP tests that aren't particularly related to the BYOArgo feature

Reviewer 2: Yeah, these are more generic, but I wanted to add regression coverage for driver and launcher (as these tests cover everything that driver and launcher are supposed to do)

Reviewer 2: @gmfrasca Please let me know if you have you think my thinking is not right here

Reviewer 3: That's fine, but what I'm getting at is that we should include a summary in the top heading of these sections to explain the reasoning for performing these tests - right now they appear to be a list of tests with no particular purpose

Reviewer 2: Good point, Done! I also changed the context file to do so for future test plan creations

| **Expected Results**  | - Pipeline submits successfully<br/>- Execution progresses normally<br/>- Results accessible through DSP interface<br/>- Logs and monitoring functional                                  |

### 2.2 Complex Pipeline Types

| Test Case ID          | TC-PF-002                                                                                                                                                                                                                                      |
|-----------------------|------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| **Test Case Summary** | Verify run of "Pipelines with artifacts" pipeline                                                                                                                                                                                              |

> | **Test Steps**        | 1. Execute pipeline - Pipelines with artifacts<br/>2. Verify each pipeline type executes correctly<br/>3. Validate artifacts, metadata, and custom configurations                                                                              |

Reviewer 3: Should include the `1. Configure DSPA with external Argo` step/pre-requisite for each of these tests (2.2 and on)

Reviewer 2: good point, I will add it

Reviewer 2: Done

| **Expected Results**  | - All pipeline types execute successfully<br/>- Custom components and packages work correctly<br/>- Retry and iteration logic functions properly<br/>- Certificate handling operates securely<br/>- Artifacts and metadata preserved correctly |

| Test Case ID          | TC-PF-003                                                                                                                                                                                                           |
|-----------------------|---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| **Test Case Summary** | Verify run of "Pipelines without artifacts" pipeline                                                                                                                                                                |

> | **Test Steps**        | 1. Execute Pipeline - Pipelines without artifacts<br/>2. Verify each pipeline type executes correctly<br/>3. Validate artifacts, metadata, and custom configurations                                                |                                                                                                                                                                                   |

Reviewer 3: Validate artifacts on "Pipelines without artifacts"?

Reviewer 2: Copy pasta issue, I have already fixed few of these test cases, still working on fixing rest (in this section)

| **Test Steps**        | <ol><li> Configure DSPA with external Argo </li><li> Execute Pipeline - Pipelines without input parameters </li><li> Verify each pipeline type executes correctly</li></ol>  |
| **Expected Results**  | - Pipeline runs successfully                                                                                                                                                 |

| Test Case ID          | TC-PF-012                                                                                                                                                                 |
|-----------------------|---------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| **Test Case Summary** | Verify run of "Pipelines with NO output artifacts, but just input artifacts" pipeline                                                                                     |
| **Test Steps**        | <ol><li> Configure DSPA with external Argo </li><li> Execute Pipeline - Pipelines with output artifacts </li><li> Verify each pipeline type executes correctly</li></ol>  |

> | **Expected Results**  | - Pipeline runs successfully<br/>- Output artifacts (like a model/trained data) are produced to S3 correctly                                                              |

Reviewer: The summary doesn't seem to match the steps and expected results.
`Verify run of "Pipelines with NO output artifacts`
`Execute Pipeline - Pipelines with output artifacts`
`Output artifacts (like a model/trained data) are produced to S3 correctly`

Reviewer 2: :doh
I will fix it

Reviewer 2: Fixed

8. **CRD Management Guidelines** - Platform-level CRD update-in-place and conflict resolution procedures
9. **Pre-existing Argo Detection Guide** - Implementation and configuration of detection mechanisms
10. **Configuration Validation Guide** - Detection and resolution of all unsupported configurations
11. **RBAC Compatibility Matrix** - Comprehensive guidelines for DSP and external Argo RBAC alignment
12. **Schema Compatibility Guide** - Workflow schema version compatibility and API compatibility matrix
13. **Co-existence Best Practices** - Detailed recommendations for running DSP and external Argo together
14. **Z-Stream Testing Strategy** - Framework for ongoing patch version compatibility validation
15. **API Compatibility Documentation** - DSP API Server to external WorkflowController compatibility guidelines
16. **Independent Lifecycle Management Guide** - Best practices for managing ODH and Argo independently
17. **Known Issues Log** - Comprehensive documentation of limitations and workarounds
18. **Final Test Report** - Executive summary with recommendations, lessons learned, and future testing strategy

## Test Implementation/Execution Phases

> ### Phase 1 (Sprint 23)

Reviewer 3: I wouldn't list the sprint names (ie "Sprint 23") these are planned for, as they are not particularly useful to the general public

Reviewer 2: Removed, but we still need to decide what goes into these sections
